add app management section (#1704)
giantswarm/docs: src/content/advanced/gitops/apps/app_sets.md

---
linkTitle: Creating and using App Sets
title: Creating and using App Sets
description: Learn how to create and use App Sets for applications deployed with GitOps.
weight: 50
menu:
  main:
    identifier: advanced-gitops-apps-app-sets
    parent: advanced-gitops-apps
last_review_date: 2023-02-10
user_questions:
  - How can I create and use App Sets in GitOps?
owner:
  - https://github.com/orgs/giantswarm/teams/team-honeybadger
---

# Creating and using App Sets

> It is often desirable to deploy a group of apps together, as a single deployment step. We call such groups "App Sets". There's nothing special about App Sets: they are not a separate API entity, but rather just a configuration pattern enabled by `kustomize` and `flux`. The purpose of such an approach can vary, but it's usually to meet the following benefits:

Reviewer: ```suggestion
It is often desirable to deploy a group of apps together, as a single deployment step. We call such groups "App Sets". There's nothing special about App Sets: they are not a separate API entity, but rather just a configuration pattern enabled by [Kustomize](https://kustomize.io/) and [Flux](https://fluxcd.io/flux/). The purpose of such an approach can vary, but it's usually to meet the following benefits:
```

- Apps within the set have strong relationship and dependency on versions and you want to be sure that the correct versions are used and deployed together.
- The configuration of apps in an App Set could be shared. By placing them together it is possible to simplify the configuration of one app by providing values known to depend on another. As a result, it is easier to deploy an App Set than each of the apps individually. That way a specialized team might deliver a pre-configured App Set so that it is easy and understandable to deploy be a less proficient end user.

## Limitations

Even though App Sets have a lot of benefits, their implementation with `kustomize` and `flux` is pretty limited. Please make sure that you read this section before deciding to implement App Sets the way they are described here.

In general, the whole problem can be summarized as "`kustomize` is not a templating engine". It can override some values, even in bulks, but it can't put the same value in any arbitrary place.

> In the common case, it is impossible to configure a variable once and use it with multiple apps. There are only two choices here: either each pair of apps shares exactly the same ConfigMap/Secret as a `config:` attribute of an App CR or they use two separate ConfigMaps. In the latter case, it is your responsibility to provide exactly the same value in both ConfigMaps/Secrets. In the former, it means that your apps share the same configuration input and must be able to handle this situation correctly: there can be no conflicting options and each shared option must be understood exactly the same by each app. In practice, it means that the apps must be prepared on the Helm chart layer to work together.

Reviewer: ```suggestion
In the common case, it is impossible to configure a variable once and use it with multiple apps. There are only two choices here: either each pair of apps shares exactly the same ConfigMap/Secret as a `config:` attribute of an App CR or they use two separate ConfigMaps. In the latter case, it is your responsibility to provide exactly the same value in both ConfigMaps/Secrets. In the former, it means that your apps share the same configuration input and must be able to handle this situation correctly: there can be no conflicting options and each shared option must be understood exactly the same by each app. In practice, it means that the apps must be prepared on the Helm chart layer to work together.

```

---
linkTitle: Creating and using App Sets
title: Creating and using App Sets
description: Learn how to create and use App Sets for applications deployed with GitOps.
weight: 50
menu:
  main:
    identifier: advanced-gitops-apps-app-sets
    parent: advanced-gitops-apps
last_review_date: 2023-02-10
user_questions:
  - How can I create and use App Sets in GitOps?
owner:
  - https://github.com/orgs/giantswarm/teams/team-honeybadger
---

# Creating and using App Sets

> It is often desireable to deploy a group of apps together, as a single deployment step. We call such groups "App Sets". There's nothing special about App Sets: they are not a separate API entity, rather just a configuration pattern enabled by `kustomize` and `flux`. The purpose of such approach can vary, but it's usually to meet the following benefits:

Reviewer: ```suggestion
It is often desirable to deploy a group of apps together, as a single deployment step. We call such groups "App Sets". There's nothing special about App Sets: they are not a separate API entity, but rather just a configuration pattern enabled by `kustomize` and `flux`. The purpose of such an approach can vary, but it's usually to meet the following benefits:
```

- Apps within the set have strong relationship and dependency on versions and you want to be sure that the correct versions are used and deployed together.

> - Configuration of one app in a set depends on configuration of another app in the same set. By placing them together it is possible to simplify configuration of one app by providing values known to depend on another. As a result, it is easier to deploy an App Set than each of the apps individually. That way a specialized team might deliver a pre-configured App Set so that it is easy and understandable to deploy be a less proficient end user.

Reviewer: ```suggestion
- The configuration of apps in an App Set could be shared. By placing them together it is possible to simplify the configuration of one app by providing values known to depend on another. As a result, it is easier to deploy an App Set than each of the apps individually. That way a specialized team might deliver a pre-configured App Set so that it is easy and understandable to deploy be a less proficient end user.
```

## Limitations

Even though App Sets have a lot of benefits, their implementation with `kustomize` and `flux` is pretty limited. Please make sure that you read this section before deciding to implement App Sets the way they are described here.

> In general, the whole problem can be summarized as "`kustomize` is not a templating engine". It can override some values, even in bulks, but in can't put the same value in any arbitrary place.

Reviewer: ```suggestion
In general, the whole problem can be summarized as "`kustomize` is not a templating engine". It can override some values, even in bulks, but it can't put the same value in any arbitrary place.
```

> In general case, it is impossible to configure a variable once and use it with multiple apps. There are only 2 choices here: either each pair of apps shares exactly the same ConfigMap/Secret as a `config:` attribute of an App CR or they use two separate ConfigMaps. In the latter case, it is your responsibility to provide exactly the same value in both ConfigMaps/Secrets. In the former, it means that your apps share the same configuration input and must be able to handle this situation correctly: there can be no conflicting options and each shared option must be understood exactly the same by each app. In practice, it means that the apps must be prepared on Helm chart layer to work together.

Reviewer: ```suggestion
In the common case, it is impossible to configure a variable once and use it with multiple apps. There are only two choices here: either each pair of apps shares exactly the same ConfigMap/Secret as a `config:` attribute of an App CR or they use two separate ConfigMaps. In the latter case, it is your responsibility to provide exactly the same value in both ConfigMaps/Secrets. In the former, it means that your apps share the same configuration input and must be able to handle this situation correctly: there can be no conflicting options and each shared option must be understood exactly the same by each app. In practice, it means that the apps must be prepared on the Helm chart layer to work together.
```

If none of these two solutions is applicable, you might want to solve the problem of bundling the apps together elsewhere. One of possible routes is to create a new umbrella Helm Chart that includes all the necessary apps as [sub-charts](https://helm.sh/docs/chart_template_guide/subcharts_and_globals/).

## Example

### App Set Template

An example of an App Set Template is available in the [gitops-template repository in "bases/app_sets/hello-web-app"](https://github.com/giantswarm/gitops-template/tree/main/bases/app_sets/hello-web-app/). This App Set assumes, that it's impossible to build a shared ConfigMap for both Apps and as such does full config override on App Set template level and override using `userConfig:` field in App Set instance.

### Using App Set

An example showing how to use an App Set is available in the [gitops-template repository in "WC_NAME/app_sets/hello-web-app-1"](https://github.com/giantswarm/gitops-template/tree/main//management-clusters/MC_NAME/organizations/ORG_NAME/workload-clusters/WC_NAME/app_sets/hello-web-app-1).

## Creating an App Set Template

> Creating an App Set Template is not much different than [creating an App Template](/advanced/gitops/apps/add_app_template/). Please check the
> docs there first, if yoo haven't already.

Reviewer: ```suggestion
Creating an App Set Template is not much different than [creating an App Template](/advanced/gitops/apps/add_app_template/). Please check the docs there first, if you haven't already.
```

## Using an App Set

Using an App Set Template is once again very similar to using a single [App Set](/advanced/gitops/apps/add_appcr/#adding-app-using-app-template). To create one, save a path that contains your desired App Set Template in the [gitops-template repository in "bases/app_sets"](https://github.com/giantswarm/gitops-template/tree/main/bases/app_sets/) directory. Then, create a new directory in the `app_sets` directory of your Working Cluster. In that directory, create a `kustomization.yaml` file based on the following pattern:

```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
buildMetadata: [originAnnotations]
configMapGenerator:
  - files:
      - values=override_config_hello_world.yaml
    name: ${cluster_name}-hello-world-user-values
generatorOptions:
  disableNameSuffixHash: true
kind: Kustomization
namespace: hello-web-team1
patchesStrategicMerge:
  - config_patch.yaml
resources:
  - ../../../../../../../../../bases/app_sets/hello-web-app
```

> Over here, we are overriding the configuration of the `hello-world` app, which was already defined in the App Set Template. Since here we're using the `userConfig` property, we don't have to override the whole config, but only YAML keys we need to change. One more important fact is that we're setting a custom Namespace for the whole deployment of an app. If you want to learn more about how config overrides work, please consult our [docs about creating apps](/advanced/gitops/apps/add_appcr/), as in general App Set is just a bundle of them.

Reviewer: ```suggestion
Over here, we are overriding the configuration of the `hello-world` app, which was already defined in the App Set Template. Since here we're using the `userConfig` property, we don't have to override the whole config, but only the YAML keys we need to change. One more important fact is that we're setting a custom Namespace for the whole deployment of an app. If you want to learn more about how config overrides work, please consult our [docs about creating apps](/advanced/gitops/apps/add_appcr/), as in general App Set is just a bundle of them.
```
